theano/gpuarray/opt.py

```diff
@@ -215,6 +256,242 @@ def apply(self, fgraph):
                     0, 'fast_run', 'fast_compile', 'merge')
 
 
+class GraphToGPU(NavigatorOptimizer):
+    """
+    Transfer the graph as a whole to GPU instead of transfering node by node.
+    """
+
+    def __init__(self, local_optimizers_all, local_optimizers_map):
+        self.local_optimizers_all = local_optimizers_all
+        self.local_optimizers_map = local_optimizers_map
+        self.failure_callback = None
+
+    def add_requirements(self, fgraph):
+        fgraph.attach_feature(toolbox.ReplaceValidate())
+
+    def apply(self, fgraph):
+        mapping = {}
+        time_opts = {}
+        node_created = {}
+        process_count = {}
+        t_topo = time.time()
+        topo = fgraph.toposort()
+        time_topo = time.time()
+        toposort_timing = time_topo - t_topo
+
+        # Building a new graph
+        # Iterating through inputs of graph
+        target = infer_context_name(*fgraph.inputs)
+        for i in fgraph.inputs:
+            # Do not move *int* scalar to the GPU.
+            target = getattr(i.tag, 'target', None)
+            if (target != 'cpu' and
+               isinstance(i.type, tensor.TensorType) and
+               (i.ndim > 0 or 'int' not in i.dtype)):
+                mapping[i] = as_gpuarray_variable(i, target)
+            else:
+                mapping[i] = i
+        for i in fgraph.variables:
+            if isinstance(i, theano.Constant):
+                mapping[i] = i
+        for node in topo:
+            for lopt in (self.local_optimizers_map.get(node.op, []) +
+                         self.local_optimizers_map.get(type(node.op), []) +
+                         self.local_optimizers_all):
+                    process_count.setdefault(lopt, 0)
+                    time_opts.setdefault(lopt, 0)
+                    node_created.setdefault(lopt, 0)
+
+        for node in topo:
+
+            if isinstance(node.op, HostFromGpu):
+                mapping[node.outputs[0]] = mapping[node.inputs[0]]
+                continue
+
+            # Move only if any of the inputs are on the GPU.
+            move_to_GPU = False
+            if any([isinstance(i, GpuArrayVariable) or
+                   isinstance(i, GpuArraySharedVariable)
+                   for i in [mapping[v] for v in node.inputs] +
+                   node.outputs]):
+
+                move_to_GPU = True
+
+            context_name = None
+            for i in [mapping[i] for i in node.inputs]:
+                target = getattr(i.tag, 'target', None)
```

Comment:
OK. I was remembering the general logic, but did not remember where it was.
If it is explicitly embedded in the graph via a `HostFromGpu` node, I guess we are OK.